packages/pymedphys_dicom/src/pymedphys_dicom/ct/extend.py

```diff
@@ -0,0 +1,180 @@
+
+    filenames = [
+        os.path.basename(filepath)
+        for filepath in filepaths
+    ]
+
+    common_prefix = os.path.commonprefix(filenames)
+
+    number_of_digits = len(str(len(dicom_datasets)))
+    new_filenames = [
+        "{}{}.dcm".format(
+            common_prefix,
+            str(dicom_dataset.InstanceNumber).zfill(number_of_digits)
+        )
+        for dicom_dataset in dicom_datasets
+    ]
+
+    new_filepaths = [
+        os.path.join(output_dir, filename)
+        for filename in new_filenames
+    ]
+
+    for dicom_dataset, filepath in zip(dicom_datasets, new_filepaths):
+        dicom_dataset.save_as(filepath)
+
+
+def extend_datasets(dicom_datasets, index_to_copy, number_of_slices, uids=None):
+    copy_slices_and_append(dicom_datasets, index_to_copy, number_of_slices)
+    refresh_instance_numbers(dicom_datasets)
+    generate_new_uids(dicom_datasets, uids=uids)
+
+
+def load_dicom_into_deque(filepaths):
+    dicom_datasets_initial_read = [
+        pydicom.dcmread(filepath, force=True)
+        for filepath in filepaths
+    ]
+
+    dicom_datasets = convert_datasets_to_deque(dicom_datasets_initial_read)
+
+    return dicom_datasets
+
+
+def convert_datasets_to_deque(datasets):
+    dicom_datasets = deque()
+
+    for dicom_dataset in sorted(datasets, key=slice_location):
+        dicom_datasets.append(dicom_dataset)
+
+    return dicom_datasets
+
+
+
+def instance_number(dicom_dataset):
+    return dicom_dataset.InstanceNumber
+
+
+def slice_location(dicom_dataset):
+    return float(dicom_dataset.SliceLocation)
+
+
+def copy_slices_and_append(dicom_datasets, index_to_copy, number_of_slices):
+    append_method = get_append_method(dicom_datasets, index_to_copy)
+    new_slice_locations = generate_new_slice_locations(
+        dicom_datasets, index_to_copy, number_of_slices)
+
+    dataset_to_copy = deepcopy(dicom_datasets[index_to_copy])
+
+    append = getattr(dicom_datasets, append_method)
+
+    for slice_location in new_slice_locations:
+        new_slice = deepcopy(dataset_to_copy)
+
+        new_slice.SliceLocation = str(slice_location)
+
+        image_position_patient_to_copy = deepcopy(
+            dicom_datasets[index_to_copy].ImagePositionPatient)
+        image_position_patient_to_copy[-1] = str(slice_location)
+        new_slice.ImagePositionPatient = image_position_patient_to_copy
+
+        append(new_slice)
+
+
+def refresh_instance_numbers(dicom_datasets):
+    for i, dicom_dataset in enumerate(dicom_datasets):
+        dicom_dataset.InstanceNumber = str(i)
+
+
+def generate_new_uids(dicom_datasets, uids=None):
+    if uids is None:
+        uids = generate_uids(len(dicom_datasets))
+
+    for dicom_dataset, uid in zip(dicom_datasets, uids):
+        dicom_dataset.SOPInstanceUID = uid
+
+
+def generate_new_slice_locations(dicom_datasets, index_to_copy, number_of_slices):
+    if index_to_copy == 0:
+        slice_diff = dicom_datasets[0].SliceLocation - \
```

Comment:
One almost insignificant point:

It might pay to use `ds.ReconstructionProtocolElementSequence.SpacingBetweenSlices` **if it exists** (it's an optional parameter).

This accounts for cases with missing slices. Not as rare as one would hope...

Reply:
Or get minimum separation across all slices?

Reply:
Yup, very fair points. Not insignificant at all.

Reply:
Could you provide me a DICOM file that has the `SpacingBetweenSlices` element on it?

Reply:
I don't actually have one. I only thought of it from my own research running into this issue for a script that force aligns a CT to the z-axis

Incidentally, I was planning to add this to `pymedphys`. The use case is that we often come across diagnostic CT scans from scanners that seem to have a very slight skew between the imaging and scanner coordinate systems (and RayStation insists upon perfect z-axis alignment - for now at least).

Reply:
I did try to find one on the internet to no avail. I'll keep looking.

Reply:
Kk, thanks @Centrus007